modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/views/table/Table.tsx

```diff
@@ -0,0 +1,477 @@
+
+// @ts-ignore
+
+import FieldsSelectorDropdown from './FieldsSelectorDropdown';
+import TableCell from './TableCell';
+import TableHeadCell from './TableHeadCell';
+
+// @ts-ignore
+
+import TableInlineAddingRow from './TableInlineAddingRow';
+import TableContext from './dnd_table/TableContext';
+
+// @ts-ignore
+
+import DndTable from './dnd_table/index';
+
+interface IField {
+	fieldName: string | [];
+	label: string;
+	mapData: Function;
+}
+interface ISchema {
+	fields: Array<IField>;
+}
+
+const getVisibleFields = ({
+	fields,
+	visibleFieldNames,
+}: {
+	fields: Array<any>;
+	visibleFieldNames: Array<string>;
+}) => {
+	const visibleFields = fields.filter(
+		({fieldName}) => visibleFieldNames[fieldName]
+	);
+
+	return visibleFields.length ? visibleFields : fields;
+};
+
+const Head = ({
+	fields,
+	items,
+	itemsActions,
+	schema,
+	selectItems,
+	selectable,
+	selectedItemsKey,
+	selectedItemsValue,
+	selectionType,
+}: {
+	fields: Array<IField>;
+	items: Array<any>;
+	itemsActions: Array<IItemsActions>;
+	schema: ISchema;
+	selectItems: Function;
+	selectable: boolean | undefined;
+	selectedItemsKey: string;
+	selectedItemsValue: any;
+	selectionType: string | undefined;
+}) => {
+	const {isFixed} = useContext(TableContext);
+
+	function handleCheckboxClick() {
+		if (selectedItemsValue.length === items.length) {
+			return selectItems([]);
+		}
+
+		return selectItems(items.map((item) => item[selectedItemsKey]));
+	}
+
+	return (
+		<DndTable.Head>
+			<DndTable.Row>
+				{selectable && (
+					<DndTable.Cell
+						className="item-selector"
+						columnName="item-selector"
+						heading
+					>
+						{!!items.length && selectionType === 'multiple' ? (
+							<ClayCheckbox
+								checked={!!selectedItemsValue.length}
+								indeterminate={
+									!!selectedItemsValue.length &&
+									items.length !== selectedItemsValue.length
+								}
+								name="table-head-selector"
+								onChange={handleCheckboxClick}
+								title={
+									items.length !== selectedItemsValue.length
+										? Liferay.Language.get('select-items')
+										: Liferay.Language.get(
+												'clear-selection'
+										  )
+								}
+							/>
+						) : null}
+					</DndTable.Cell>
+				)}
+
+				{fields.map((field) => (
+					<TableHeadCell {...field} key={field.label} />
+				))}
+
+				<DndTable.Cell
+					className="item-actions"
+					columnName="item-actions"
+					defaultWidth={0}
+					heading
+				>
+					<FieldsSelectorDropdown fields={schema.fields} />
+
+					{!isFixed &&
+						itemsActions?.length === 1 &&
+						!itemsActions[0].icon &&
+						itemsActions[0].label && (
+							<LinkOrButton
+								className="btn btn-secondary btn-sm"
+								href="#"
+								monospaced={false}
+								style={{
+									'visibility': 'hidden',
+									'white-space': 'nowrap',
+								}}
+							>
+								{itemsActions[0].label}
+							</LinkOrButton>
+						)}
+				</DndTable.Cell>
```

Comment:
This is the main change in this PR. The solution works like this:
1. On first render:
   - Widths are not set to `styles` yet. `isFixed` is the flag that is set once widths are set. 
   - We temporarily render a button, containing the label, in the header. 
   - The button is not visible, but still keeps its layout dimensions, thanks to how `visibility: 'hidden'` works.
   - We set the `defaultWidth` to 0, so the width is determined by the content of the cell.
2. On second render:
   - Widths are set, `isFixed` is `true`
   - We no longer show button in DOM

Notes:
- Although it's still possible that customers' styles break this, with very specific CSS selectors, it's not very likely.  This is probably good enough.
- There is a case that this solution does not solve, where items have different sets of actions, and some of them are with one actions. That case, I'm fairly sure, is not possible to solve without glitching table on data arrival. There is a workaround for it though, the `uniformActionsDisplay` prop, that renders ellipsis on everything.